lib/media-controls-impl.js

```diff
@@ -0,0 +1,524 @@
+
+          var direction = null;
+          if (e.target === this.els.seekForward) {
+            direction = 1;
+          } else if (e.target === this.els.seekBackward) {
+            direction = -1;
+          } else {
+            return;
+          }
+
+          var offset = direction * 10;
+          this.seekBy(this.mediaPlayer.currentTime + offset);
+
+          // Begin the "longpress" movement after a one second delay.
+          var self = this;
+          this.intervalId = window.setInterval(function() {
+              self.seekBy(self.mediaPlayer.currentTime + offset);
+            }, 1000);
+          break;
+      }
+      break;
+
+      case 'touchend':
+      case 'mouseup':
+        //
+        // If ending a long-press forward or backward, clear timer
+        //
+        if (this.intervalId) {
+           window.clearInterval(this.intervalId);
+           this.intervalId = null;
+        }
+        else if (this.dragging) {
+          // If ending a movement of the slider, end slider movement
+          this.handleSliderMoveEnd();
+        }
+
+        if (e.type === 'mouseup') {
+          // Don't listen for mousemove and mouseup until we get a mousedown
+          window.removeEventListener('mousemove', this, true);
+          window.removeEventListener('mouseup', this, true);
+          this.mouseEventHandlerRegistered = false;
+        }
+        break;
+
+    case 'loadedmetadata':
+      //
+      // Metadata has been loaded, now we can set the duration of the media
+      //
+      this.els.durationText.textContent = this.formatTime(this.mediaPlayer.duration);
+      break;
+
+    case 'play':
+      //
+      // Media is playing, display 'paused' icon
+      //
+      this.els.play.classList.remove('paused');
+
+      // Update l10n-id for the benefit of the screen reader
+      this.els.play.setAttribute('data-l10n-id', 'playbackPlay');
+      break;
+
+    case 'pause':
+      //
+      // Media is paused, display 'play' icon
+      //
+      this.els.play.classList.add('paused');
+
+      // Update l10n-id for the benefit of the screen reader
+      this.els.play.setAttribute('data-l10n-id', 'playbackPause');
+
+      // If the paused event comes when the media is at the end,
+      // set our 'played-till-end' flag. The one exception is when
+      // the paused event comes from pausing the media when the
+      // forward button was used to seek to the end while the media
+      // was playing. In this case, we don't consider the media being
+      // played until the end.
+      if (this.mediaPlayer.currentTime === this.mediaPlayer.duration) {
+        if (this.pausedAtEndWhilePlaying) {
+          this.pausedAtEndWhilePlaying = false;
+        }
+        else {
+          this.playedUntilEnd = true;
+        }
+      }
+      break;
+
+    case 'timeupdate':
+      //
+      // Update the progress bar and play head as the video plays
+      //
+      if (!this.mediaControlsElement.hidden) {
```

Comment:
I assume that this is an optimization so we don't do unnecessary updates if the element is not visible.  I think I've see code in the music app that looks at document visiblity, so if the music app is in the background the update isn't done.

Doing somethign similar here is a good idea, but I'm not sure that .hidden is the best way.  I think that setting .hidden on an element is like display:none. It just goes away. In practice apps are more likely to want to do some kind of opacity fade or animated transform to hide the controls, and are unlikely to hide the controls with .hidden (especially since this is not a document feature!)

Ideally you want to only update the controls when the user can see them and not (for example) when they've scrolled offscreen.  I'm not sure how you can determine this. And you'd also need some kind of notification when the controls are visible again so you could update them even if the media had ended and was no longer sending timeupdate events.

For now, I think you should update the controls unconditionally. As it stands, if you make the controls .hidden, then play to the end of the media, then make them visible again, the thumb will be in the wrong place, which seems like a pretty bad bug.

But maybe it is worth creating a TODO file that lists this visibility optimization as something we should add in the future.